### PR TITLE
drivers: tmp_108: ti tmp108 support

### PR DESCRIPTION
**Pull request by catch-twenty-two**
This adds support for the TI TMP108 temperature sensor. This includes over/under temp interrupt support as well as one shot, continuous conversion and power down modes.

Signed-off-by: Jimmy Johnson <catch22@fastmail.net>

### CODE REVIEW

File: drivers/sensor/CMakeLists.txt
Comment (MaureenHelm) [inline drivers/sensor/CMakeLists.txt:89]: Please replace spaces with tabs to be consistent with the rest of the file

File: drivers/sensor/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/Kconfig:215]: Please move this between tmp007 and tmp112

File: drivers/sensor/tmp108/tmp108.c
Comment (MaureenHelm) [inline drivers/sensor/tmp108/tmp108.c:31]: Store the I2C address in a const device configuration struct to enable multiple instances of this driver. You can also use `struct i2c_dt_spec` and associated helper functions like `i2c_burst_read_dt`. See #37359 for details

File: drivers/sensor/tmp108/tmp108.c
Comment (MaureenHelm) [inline drivers/sensor/tmp108/tmp108.c:301]: Similar to i2c, you can use `struct gpio_dt_spec` and associated helper functions.

File: drivers/sensor/tmp108/tmp108.c
Comment (MaureenHelm) [inline drivers/sensor/tmp108/tmp108.c:350]: Please update to support multiple instances by using `DT_INST_FOREACH_STATUS_OKAY`. See https://docs.zephyrproject.org/latest/guides/dts/howtos.html#write-device-drivers-using-devicetree-apis for details.

File: drivers/sensor/tmp108/tmp108.h
Comment (MaureenHelm) [inline drivers/sensor/tmp108/tmp108.h:28]: `BIT(n)` would be useful here. From `#include <sys/util.h>`

File: samples/sensor/tmp108/CMakeLists.txt
Comment (MaureenHelm) [inline samples/sensor/tmp108/CMakeLists.txt:3]: Please move the new sample to a separate commit

File: samples/sensor/tmp108/README.txt
Comment (MaureenHelm) [inline samples/sensor/tmp108/README.txt:1]: This file should be in .RST format

**Review by MaureenHelm — DISMISSED** [review]
Please add this sensor to `tests/drivers/build_all/sensor` and fix the failing checks (compliance, scancode, buildkite). 

Thank you for the contribution!

**Review by MaureenHelm — COMMENTED** [review]
> For my sample code for the tmp108, should I submit that after this pr is okayed and pulled in? Not sure how to do that separately from the driver without having it fail the tests.

You can push the second commit to the same branch, which will update your PR to contain two commits (one for the driver and one for the sample application)

> Please add this sensor to tests/drivers/build_all/sensor and fix the failing checks (compliance, scancode, buildkite).

`tests/drivers/build_all/sensor/prj.conf` needs to be updated in addition to the dts, and the failing checks still need to be addressed.

**Comment by catch-twenty-two** [comment]
@MaureenHelm @mbolivar-nordic thanks for the reviews! Hopefully I have addressed everything as expected. For some reason my commits came in backwards here (driver after sample) where as in my repository they are shown in the opposite direction.

File: drivers/sensor/tmp108/CMakeLists.txt
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/CMakeLists.txt]: Done

File: drivers/sensor/CMakeLists.txt
Comment (catch-twenty-two) [reply drivers/sensor/CMakeLists.txt]: Ok, Should be done now

File: samples/sensor/tmp108/README.rst
Comment (mbolivar-nordic) [inline samples/sensor/tmp108/README.rst:1]: This page's content doesn't fit the usual template and doesn't build properly into HTML docs. Please revise.

You might look at samples/sensor/dht/README.rst for an example of how this is done elsewhere -- although it's up to @MaureenHelm as sensor maintainer to declare what the right template is for sensor sample documentation.

See https://docs.zephyrproject.org/latest/guides/docs/index.html#running-the-documentation-processors for instructions on building a docs change into HTML locally.

File: samples/sensor/tmp108/src/main.c
Comment (mbolivar-nordic) [inline samples/sensor/tmp108/src/main.c:149]: This is not enough; you need to check if the sensor device was initialized correctly using `device_is_ready` here too.

**Review by mbolivar-nordic — DISMISSED** [review]
Thanks for the updates. I noticed a couple of additional things on a second pass; see inline comments.

The commit logs also need some changes to fit with the project policy:

1. the convention is for the shortlog to say `subsystem: the-change-made-to-the-subsystem` (this is one reason why the compliance check is failing)
2. zephyr requires at least one additional line of detail beyond the shortlog and the signed-off-by (this is another)

Beyond that,

3. typos, whitespace, and line length checks are failing (see compliance)

For 1., instead of `Adding ti tmp108 sensor support`, use something like `sensor: add ti tmp108 driver`. And instead of `adding ti tmp108 sensor sample`, do something like `samples: add ti tmp108 sample`.

(If unsure about the `subsystem:` prefix, just run `git log` on the directory or files you are editing and follow along with the crowd.)

For 2., please add some extra context about each patch.

This is what the failed compliance check in the CI results below is complaining about when it says:

```
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "adding ti tmp108 sensor sample"
1: UC6 Body has no content, should at least have 1 line.
```

Please address these too.

**Comment by catch-twenty-two** [comment]
@mbolivar-nordic thanks, I have applied the updates you suggested.

**Comment by catch-twenty-two** [comment]
> Thanks for the updates. I noticed a couple of additional things on a second pass; see inline comments.
> 
> The commit logs also need some changes to fit with the project policy:
> 
> 1. the convention is for the shortlog to say `subsystem: the-change-made-to-the-subsystem` (this is one reason why the compliance check is failing)
> 2. zephyr requires at least one additional line of detail beyond the shortlog and the signed-off-by (this is another)
> 
> Beyond that,
> 
> 1. typos, whitespace, and line length checks are failing (see compliance)
> 
> For 1., instead of `Adding ti tmp108 sensor support`, use something like `sensor: add ti tmp108 driver`. And instead of `adding ti tmp108 sensor sample`, do something like `samples: add ti tmp108 sample`.
> 
> (If unsure about the `subsystem:` prefix, just run `git log` on the directory or files you are editing and follow along with the crowd.)
> 
> For 2., please add some extra context about each patch.
> 
> This is what the failed compliance check in the CI results below is complaining about when it says:
> 
> ```
> 1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "adding ti tmp108 sensor sample"
> 1: UC6 Body has no content, should at least have 1 line.
> ```
> 
> Please address these too.

Addressed this, thanks!

**Comment by catch-twenty-two** [comment]
@MaureenHelm or @mbolivar it looks like all the tests are finally passing, what are next steps?

**Comment by cfriedt** [comment]
@MaureenHelm - PTAL when you can

**Review by gmarull — CHANGES_REQUESTED** [review]
There are a few common issues that should be addressed (style, error propagation), as well as some other questions I raised.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Agree with @gmarull 's comments

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Some of the changes introducing the sample application slipped into the first commit (`drivers: sensor: add ti tmp108 driver support`). Please move them to the second commit (`sample: sensor: tmp108: Add ti tmp108 sample`)

**Comment by carlescufi** [comment]
@gmarull @MaureenHelm could you take another look?

File: drivers/sensor/tmp108/tmp108.c
Comment (mbolivar-nordic) [inline drivers/sensor/tmp108/tmp108.c:395]: Since alert-gpios is an optional property, you need to use GPIO_DT_SPEC_INST_GET_OR here instead.

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks, I confirm that my existing DT and driver model comments are all addressed. I found another bug, otherwise this LGTM from that perspective.

**Comment by catch-twenty-two** [comment]
@mbolivar-nordic or @MaureenHelm Is there a recommended way to rollback the sample changes to get to the previous (driver) commit, change that, then reapply the sample changes? It's a lot of work  to make specific changes to completely separate commits on the same branch, especially since I have had to do this multiple times.

**Comment by mbolivar-nordic** [comment]
>  Is there a recommended way to rollback the sample changes to get to the previous (driver) commit, change that, then reapply and make changes to the sample?

Use an interactive git rebase.

**Comment by mbolivar-nordic** [comment]
See e.g. step 15 in https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow, although that example is more complicated than I would like.

A simpler example:

```
git checkout add-ti-tmp108-support
git rebase -i HEAD~~
# change the driver commit from 'pick' to 'edit' in the rebase TODO file, save and exit
# edit the driver, and commit it
git rebase --continue
```

**Comment by catch-twenty-two** [comment]
> See e.g. step 15 in https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow, although that example is more complicated than I would like.
> 
> A simpler example:
> 
> ```
> git checkout add-ti-tmp108-support
> git rebase -i HEAD~~
> # change the driver commit from 'pick' to 'edit' in the rebase TODO file, save and exit
> # edit the driver, and commit it
> git rebase --continue
> ```

Thanks! Ill give that a shot.

**Review by MaureenHelm — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:27]: unused define

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:29]: by convention: please move this to the top, after includes

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:25]: nit: unneeded blank lines

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:39]: redundant 0 initialization

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:42]: unneded blank line (and others)

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [reply drivers/sensor/tmp108/tmp108.c:56]: please, check `byteorder.h` API

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:111]: please, validate supported channels here (SENSOR_CHAN_ALL and SENSOR_CHAN_AMBIENT_TEMP)

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:161]: SENSOR_CHAN_ALL should not be validated here

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:165]: nit: add `U` suffix to all unsigned constants `>> 4U`

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:180]: fits into single line
```
	if (chan != SENSOR_CHAN_AMBIENT_TEMP && chan != SENSOR_CHAN_ALL) {
		return -ENOTSUP;
	}
```

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:184]: why `int` cast here?

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:215]: fits into single line
```
	if (chan != SENSOR_CHAN_AMBIENT_TEMP && chan != SENSOR_CHAN_ALL) {
		return -ENOTSUP;
	}
```

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:198]: this function will return 0 even if attr is not supported

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:333]: redundant 0 init

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [inline drivers/sensor/tmp108/tmp108.c:391]: `TMP108_DEFINE`?

File: include/drivers/sensor/tmp108.h
Comment (gmarull) [inline include/drivers/sensor/tmp108.h:37]: `U` suffix

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/sensor/tmp108/tmp108.c
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c:56]: please see my above comment.

File: drivers/sensor/tmp108/tmp108.c
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c:111]: I'm still not sure when sensor channel all should be used, can you point me to some documentation?

Comment (gmarull) [reply drivers/sensor/tmp108/tmp108.c:111]: see for example https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/lm77/lm77.c#L272

ref. https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/sensor.h#L513

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [reply drivers/sensor/tmp108/tmp108.c:56]:
```c
tx_buf[0] = reg;
sys_put_be16(val, &tx_buf[1]);
```

File: drivers/sensor/tmp108/tmp108.c
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c:391]: I'm not sure what you are referring to here?

**Comment by catch-twenty-two** [comment]
Hi @gmarull thanks for the input, I was able to make the changes requested, I just had question above.

File: drivers/sensor/tmp108/tmp108.c
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c:184]: I'm not sure, this might have been left over from testing.

File: drivers/sensor/tmp108/tmp108.c
Comment (gmarull) [reply drivers/sensor/tmp108/tmp108.c:391]: I was suggesting a name change, `INST` doesn't describe the purpose of the macro

**Review by gmarull — CHANGES_REQUESTED** [review]
Please check my comments, you may need to review your editor settings

File: drivers/sensor/tmp108/tmp108.c
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c:184]: since this includes public attributes that are not listed in the sensor attr enum, I am getting warnings, if I don't cast this as an int example:

```
/home/jimmy/zephyrupstream/zephyr/drivers/sensor/tmp108/tmp108.c:271:2: warning: case value '15' not in enumerated type 'enum sensor_attribute' [-Wswitch]
  271 |  case SENSOR_ATTR_SHUTDOWN_MODE:

```

File: drivers/sensor/tmp108/tmp108.h
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.h]: I ran uncrusitfy and tried to fix up the inconsistencies.

File: include/drivers/sensor/tmp108.h
Comment (gmarull) [inline include/drivers/sensor/tmp108.h:30]: please, namespace these custom attributes (see e.g. sgp40)

File: samples/sensor/tmp108/src/main.c
Comment (gmarull) [inline samples/sensor/tmp108/src/main.c:24]: use Kconfig for this, e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/sensor/sgp40_sht4x/Kconfig

File: samples/sensor/tmp108/src/main.c
Comment (gmarull) [inline samples/sensor/tmp108/src/main.c:138]: second printf will be executen, even on error

**Review by gmarull — CHANGES_REQUESTED** [review]
There are style issues that have not been fixed

**Comment by catch-twenty-two** [comment]
> There are style issues that have not been fixed

Is there a style guide for this? I would have thought Uncrusitfy paired with the checkpatch perl script would have sufficed,  if there isn't I would suggest that the Uncrustify script be updated. Manually changing a style to conform to every nuance is very difficult and subjective since it can vary greatly from project to project and person to person. (e.g. a persons preference for grouping functionality in a routine with spacing between calls)

**Comment by gmarull** [comment]
> > There are style issues that have not been fixed
> 
> Is there a style guide for this? I would have thought Uncrusitfy paired with the checkpatch perl script would have sufficed, if there isn't I would suggest that the Uncrustify script be updated. Manually changing a style to conform to every nuance is very difficult and subjective since it can vary greatly from project to project and person to person. (e.g. a persons preference for grouping functionality in a routine with spacing between calls)

https://docs.zephyrproject.org/latest/contribute/index.html#coding-style
https://www.kernel.org/doc/html/latest/process/coding-style.html

File: dts/bindings/sensor/ti,tmp108.yaml
Comment (teburd) [inline dts/bindings/sensor/ti,tmp108.yaml:1]: Is this really a linaro created file?

**Review by teburd — APPROVED** [review]
Looks good to me. I'd honestly suggest filing an issue on uncrustify and the docs here if you ran it and it still didn't conform. The docs suggest using it, I've done just that as well, and yet folks still note that it doesn't conform to the style guide which is I agree a confusing thing to run into. Especially as a new contributor.

File: dts/bindings/sensor/ti,tmp108.yaml
Comment (catch-twenty-two) [reply dts/bindings/sensor/ti,tmp108.yaml]: fixed

**Comment by catch-twenty-two** [comment]
> Looks good to me. I'd honestly suggest filing an issue on uncrustify and the docs here if you ran it and it still didn't conform. The docs suggest using it, I've done just that as well, and yet folks still note that it doesn't conform to the style guide which is I agree a confusing thing to run into. Especially as a new contributor.

Agreed, we are preparing to submit more drivers and updates to drivers, I'd hate for things to blocked because of stylistic issues that can't be address without multiple updates to prs.

**Comment by gmarull** [comment]
> > Looks good to me. I'd honestly suggest filing an issue on uncrustify and the docs here if you ran it and it still didn't conform. The docs suggest using it, I've done just that as well, and yet folks still note that it doesn't conform to the style guide which is I agree a confusing thing to run into. Especially as a new contributor.
> 
> Agreed, we are preparing to submit more drivers and updates to drivers, I'd hate for things to blocked because of stylistic issues that can't be address without multiple updates to prs.

I agree that it's annoying, and unfortunately tooling like uncrustify seem to be outdated or not working as expected (there are some compliance issues that are likely related to that). You may try with clang-format, from my experience it is not perfect but does a decent job.

**Comment by MaureenHelm** [comment]
@gmarull have your requested changes been addressed?

**Comment by gmarull** [comment]
> @gmarull have your requested changes been addressed?

I don't see any changes related to my last comments, @catch-twenty-two can you please check?

**Comment by catch-twenty-two** [comment]
I haven't had time to look into clang and haven't used that before so I'll probably wont get around to it before the new year. I believe the only problems left are style issues.

**Comment by catch-twenty-two** [comment]
@gmarull i took another quick look today, it seems like uncrusitfy has matched the formatting of the other files in the project which do not strictly use only spaces but a mixture with 8 width tabs preceding spaces to align. @MaureenHelm fyi

**Review by gmarull — APPROVED** [review]
I guess it's ok to move on. I'll try to spend some time digging into the status of formatters and see if they can be improved.

**Comment by catch-twenty-two** [comment]
> I guess it's ok to move on. I'll try to spend some time digging into the status of formatters and see if they can be improved.

Thanks, just fyi I have opened an issue for this here: https://github.com/zephyrproject-rtos/zephyr/issues/41281

**Pull request merged**
